Sources/MapboxCoreNavigation/RerouteController.swift

```diff
@@ -64,20 +90,55 @@ extension RerouteController: RerouteObserver {
     }
 
     func onRerouteDetected(forRouteRequest routeRequest: String) {
-        // TODO: fill with Native RerouteController integration
-        defaultRerouteController.cancel()
+        isCancelled = false
+        recentRouteResponse = nil
+        guard reroutesProactively else { return }
+        delegate?.rerouteControllerDidDetectReroute(self)
     }
 
     func onRerouteReceived(forRouteResponse routeResponse: String, routeRequest: String, origin: RouterOrigin) {
-        // TODO: fill with Native RerouteController integration
+        guard reroutesProactively else { return }
+        
+        guard let decodedRequest = Self.decode(routeRequest: routeRequest) else {
+            delegate?.rerouteControllerDidFailToReroute(self, with: DirectionsError.invalidResponse(nil))
+            return
+        }
+        
+        if let recentRouteResponse = recentRouteResponse,
+           decodedRequest.routeOptions == recentRouteResponse.options {
+            delegate?.rerouteControllerDidRecieveReroute(self,
+                                                         response: recentRouteResponse.response,
+                                                         options: recentRouteResponse.options)
+        } else {
+            guard let decodedResponse = Self.decode(routeResponse: routeResponse,
+                                                    routeOptions: decodedRequest.routeOptions,
+                                                    credentials: decodedRequest.credentials) else {
+                delegate?.rerouteControllerDidFailToReroute(self, with: DirectionsError.invalidResponse(nil))
+                return
+            }
+            
+            delegate?.rerouteControllerDidRecieveReroute(self,
```

Comment:
Do we need to cache `recentRouteResponse` here?

Reply:
No. `recentRouteResponse` is cached when SDK uses custom `routingProvider` for a request to save extra encoding/decoding operation. Thus `onRerouteReceived` is the final consumer of the cached value and there is no sense to keep it further. So it is even better to nillify `recentRouteResponse` at the end of this method to free some memory.